tests/unit/completion/test_models.py

```diff
@@ -0,0 +1,386 @@
+        ]),
+        ('1', [
+            ('1/1', 'https://wiki.archlinux.org', 'ArchWiki'),
+        ])
+    ]
+
+
+def test_tab_completion_delete(stubs, qtbot, app_stub, win_registry,
+                               tabbed_browser_stubs):
+    """Verify closing a tab by deleting it from the completion widget."""
+    tabbed_browser_stubs[0].tabs = [
+        stubs.FakeWebView(QUrl('https://github.com'), 'GitHub', 0),
+        stubs.FakeWebView(QUrl('https://wikipedia.org'), 'Wikipedia', 1),
+        stubs.FakeWebView(QUrl('https://duckduckgo.com'), 'DuckDuckGo', 2)
+    ]
+    tabbed_browser_stubs[1].tabs = [
+        stubs.FakeWebView(QUrl('https://wiki.archlinux.org'), 'ArchWiki', 0),
+    ]
+    model = miscmodels.TabCompletionModel()
+    view = _mock_view_index(model, 0, 1)
+    model.delete_cur_item(view)
+    actual = [tab.url() for tab in tabbed_browser_stubs[0].tabs]
+    assert actual == [QUrl('https://github.com'),
+                      QUrl('https://duckduckgo.com')]
+
+
+def test_setting_section_completion(monkeypatch, stubs):
+    module = 'qutebrowser.completion.models.configmodel'
+    _patch_configdata(monkeypatch, stubs, module + '.configdata.DATA')
+    _patch_config_section_desc(monkeypatch, stubs,
+                               module + '.configdata.SECTION_DESC')
+    actual = _get_completions(configmodel.SettingSectionCompletionModel())
+    assert actual == [
+        ("Sections", [
+            ('general', 'General/miscellaneous options.', ''),
+            ('ui', 'General options related to the user interface.', ''),
+        ])
+    ]
+
+
+def test_setting_option_completion(monkeypatch, stubs, config_stub):
+    module = 'qutebrowser.completion.models.configmodel'
+    _patch_configdata(monkeypatch, stubs, module + '.configdata.DATA')
+    config_stub.data = {'ui': {'gesture': 'off',
+                               'mind': 'on',
+                               'voice': 'sometimes'}}
+    actual = _get_completions(configmodel.SettingOptionCompletionModel('ui'))
+    assert actual == [
+        ("ui", [
+            ('gesture', 'Waggle your hands to control qutebrowser', 'off'),
+            ('mind', 'Enable mind-control ui (experimental)', 'on'),
+            ('voice', 'Whether to respond to voice commands', 'sometimes'),
+        ])
+    ]
+
+
+def test_setting_value_completion(monkeypatch, stubs, config_stub):
+    module = 'qutebrowser.completion.models.configmodel'
+    _patch_configdata(monkeypatch, stubs, module + '.configdata.DATA')
+    config_stub.data = {'general': {'volume': '0'}}
+    model = configmodel.SettingValueCompletionModel('general', 'volume')
+    actual = _get_completions(model)
+    assert actual == [
+        ("Current/Default", [
+            ('0', 'Current value', ''),
+            ('11', 'Default value', ''),
+        ]),
+        ("Completions", [
+            ('0', '', ''),
+            ('11', '', ''),
+        ])
+    ]
+
+
+def _get_completions(model):
```

Comment:
I think all those utility functions should be at the top of the file rather than the bottom. Makes it easier to read the file from top to bottom :wink: